### PR TITLE
feat: kyverno policies

### PR DESCRIPTION
**Pull request by srodenhuis**
This MVP includes:

- adding Kyverno as a core app with 2 deployment profiles (DevTest and Prod)
- offer 5 policy profiles with (to start with) 25 policies
- policies are per team
- teams can customize policies, deactivate them or change the default action (enforce, audit)
- admins can block teams from changing policies

### CODE REVIEW

File: charts/team-ns/templates/policies/best-practice/disallow-latest-tag.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      - resources:
+          kinds:
+          - Pod
+    skipBackgroundRequests: true
+    validate:
+      message: "Using a mutable image tag e.g. 'latest' is not allowed."
+      pattern:
+        spec:
+          containers:
+          - image: "!*:latest"
```
Comment (merll) [inline charts/team-ns/templates/policies/best-practice/disallow-latest-tag.yaml:51]: Is the exemption using the label `policy.otomi.io/ignore` also considered? We currently use a `latest` tag image on buildpacks for example.

**Review by CasLubbers — CHANGES_REQUESTED** [review]
Some things I found which can be removed I think:
- The file `bin/build-gatekeeper-artifacts.sh` can also be removed right?
- In the README.md line 133 `OPA/Gatekeeper...` can be removed right?
- The policies directory can also be removed right? If so the project structure in docs/development.md should be adjusted corectly.
- The checkPolicies command in the Otomi CLI can also be removed right? This requires changes in `index.ts`, `binzx/README.md`, `check-policies.ts`, `tests.ts`
- The file docs/policies.md should be updated or removed right?
- In the file src/cmd/destroy.ts on line 105 `gatekeeper.sh` can be removed right?
- In the file src/cmd/test.ts on line 21 `  if (!values?.apps.gatekeeper!.disableValidatingWebhook) await checkPolicies()` can be removed right?
- In the file tests/integration/upgrade.yaml on line 18 `gatekeeper` can be removed right?

**Comment by CasLubbers** [comment]
One more question, maybe I cannot find it because the PR is so big. But what are the default policies we set for the team-admin namespace?

**Comment by merll** [comment]
I have attempted but failed to deploy this branch. Some pods fail to start:

kyverno/kyverno-admission-controller
```
2024-04-17T08:13:09Z    ERROR    klog    cache/reflector.go:147    k8s.io/client-go@v0.29.3/tools/cache/reflector.go:229: Failed to watch *v2alpha1.GlobalContextEntry: failed to list *v2alpha1.GlobalContextEntry: the server could not find the requested resource (get globalcontextentries.kyverno.io)
2024-04-17T08:13:31Z    INFO    klog    cache/reflector.go:539    k8s.io/client-go@v0.29.3/tools/cache/reflector.go:229: failed to list *v2alpha1.GlobalContextEntry: the server could not find the requested resource (get globalcontextentries.kyverno.io)
```

kyverno/kyverno-cleanup-controller
```
2024-04-17T08:11:43Z    ERROR    setup    cleanup-controller/main.go:124    sanity checks failed    {"error": "failed to check CRD cleanuppolicies.kyverno.io is installed: customresourcedefinitions.apiextensions.k8s.io \"cleanuppolicies.kyverno.io\" is forbidden: User \"system:serviceaccount:kyverno:kyverno-cleanup-controller\" cannot get resource \"customresourcedefinitions\" in API group \"apiextensions.k8s.io\" at the cluster scope; failed to check CRD clustercleanuppolicies.kyverno.io is installed: customresourcedefinitions.apiextensions.k8s.io \"clustercleanuppolicies.kyverno.io\" is forbidden: User \"system:serviceaccount:kyverno:kyverno-cleanup-controller\" cannot get resource \"customresourcedefinitions\" in API group \"apiextensions.k8s.io\" at the cluster scope", "errorCauses": [{"error": "failed to check CRD cleanuppolicies.kyverno.io is installed: customresourcedefinitions.apiextensions.k8s.io \"cleanuppolicies.kyverno.io\" is forbidden: User \"system:serviceaccount:kyverno:kyverno-cleanup-controller\" cannot get resource \"customresourcedefinitions\" in API group \"apiextensions.k8s.io\" at the cluster scope"}, {"error": "failed to check CRD clustercleanuppolicies.kyverno.io is installed: customresourcedefinitions.apiextensions.k8s.io \"clustercleanuppolicies.kyverno.io\" is forbidden: User \"system:serviceaccount:kyverno:kyverno-cleanup-controller\" cannot get resource \"customresourcedefinitions\" in API group \"apiextensions.k8s.io\" at the cluster scope"}]}
```

Any ideas?

File: values-schema.yaml [hunk 2]
```diff
@@ -2387,6 +2358,30 @@ properties:
...
                 $ref: '#/definitions/resources'
+      kyverno:
+        additionalProperties: false
+        properties:
+          enabled:
+            type: boolean
+            default: false
+          _rawValues:
+            $ref: '#/definitions/rawValues'
+          mode:
```
Comment (j-zimnowoda) [inline values-schema.yaml:2369]: Is mode solely for setting HA ? 
If so, could you rename it to ha, with type boolean?

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
The previous implementation gave us better control over integrated apps.
This one validates team resources against policies only.
Although, I see a lot of benefits in terms of reporting, I do not understand why did you abandon the policy checks for platform. Could you elaborate on that?

**Review by j-zimnowoda — APPROVED** [review]

**Comment by j-zimnowoda** [comment]
> Some things I found which can be removed I think:
> 
> * The file `bin/build-gatekeeper-artifacts.sh` can also be removed right?
> * In the README.md line 133 `OPA/Gatekeeper...` can be removed right?
> * The policies directory can also be removed right? If so the project structure in docs/development.md should be adjusted corectly.
> * The checkPolicies command in the Otomi CLI can also be removed right? This requires changes in `index.ts`, `binzx/README.md`, `check-policies.ts`, `tests.ts`
> * The file docs/policies.md should be updated or removed right?
> * In the file src/cmd/destroy.ts on line 105 `gatekeeper.sh` can be removed right?
> * In the file src/cmd/test.ts on line 21 `  if (!values?.apps.gatekeeper!.disableValidatingWebhook) await checkPolicies()` can be removed right?
> * In the file tests/integration/upgrade.yaml on line 18 `gatekeeper` can be removed right?

Good points! We are going to perform further refactoring after this PR I merged.

**Pull request merged**